trino/dbapi.py

```diff
@@ -301,52 +301,35 @@ def setinputsizes(self, sizes):
     def setoutputsize(self, size, column):
         raise trino.exceptions.NotSupportedError
 
-    def _prepare_statement(self, operation, statement_name):
-        """
-        Prepends the given `operation` with "PREPARE <statement_name> FROM" and
-        executes as a prepare statement.
-
-        :param operation: sql to be executed.
-        :param statement_name: name that will be assigned to the prepare
-            statement.
+    def _new_request_with_session_from(self, request):
```

Comment:
This is better than `copy.deepcopy` but I still hate it.

I think the way we try to make prepared statement work needs improvement.
Currently it's modelled as:

- A `Connection` creates a `ClientSession` - makes total sense - all persistent information for a connection is maintained using the session (roles, prepared statements, session properties, transactions)
- A `Cursor` is created from `Connection` which either gets it's own `TrinoRequest` or re-uses one from an existing `transaction`. The `TrinoRequest` inherits a reference to the `Connection`'s `ClientSession`. This doesn't make too much sense. It assumes a Cursor will only execute a query at a time (reasonable) but also leaks HTTP protocol information to the `Cursor`.
- `Cursor.execute(sql, params)` is called which creates a `TrinoQuery` using the `Cursor`'s `TrinoRequest`.
- Any operation on the cursor then affects the `TrinoRequest` - which means we should not share `TrinoRequest` objects.

These are the current rules, we break them very badly:

- In `Cursor.execute(sql, params)` if params are provided we need to execute `PREPARE statement FROM sql`
- To execute something we need a `TrinoRequest` - so we re-use the one from current Cursor - bad bad because any other future query or running query would get affected
- We workaround this by deepcopy. This copies too much and means that the `ClientSession` is no longer shared.
- We then workaround this by creating new request but re-using the `ClientSession` from old one (to persist headers etc.) What if we miss other things that need to be re-used? Things will break.
- We then execute our PREPARE with copied request, EXECUTE using original request and then DEALLOCATE using copied request.

I'll create an issue with some pseudocode class outline mentioning what refactoring I want to do to improve this - we can discuss other possible improvements there and improve what is honestly a very organically grown code which does not have enough structure to safeguard us from doing mistakes or introducing changes with confidence that we only affect what we changed.